websocket_test.go

```diff
@@ -52,3 +52,286 @@ func TestWebsocketNil(t *testing.T) {
 
 	ws.chain.assertFailed(t)
 }
+
+func TestWebsocketExpect(t *testing.T) {
+
+	t.Run("websocket is closed", func(t *testing.T) {
+		r := newStringReporter()
+		c := &Websocket{
+			chain:    makeChain(r),
+			conn:     &websocket.Conn{},
+			isClosed: true,
+		}
+
+		c.Expect()
+
+		if !r.reported {
+			t.Errorf("Websocket.CloseWithText() error message not reported")
+		}
+
+		want := "\nunexpected read from closed WebSocket connection"
+
+		if got := r.msg; got != want {
+			t.Errorf("Websocket.CloseWithText() = %v, want %v", got, want)
+		}
+	})
+}
+
+func TestWebsocketCheckUnusable(t *testing.T) {
+	type fields struct {
+		conn     *websocket.Conn
+		isClosed bool
+	}
+	type args struct {
+		reporter *stringReporter
+		where    string
+	}
+	tests := []struct {
+		name     string
+		fields   fields
+		args     args
+		want     bool
+		reported bool
+		wantMsg  string
+	}{
+		{
+			name: "conn is nil",
+			fields: fields{
+				conn:     nil,
+				isClosed: false,
+			},
+			args: args{
+				reporter: newStringReporter(),
+				where:    "Close",
+			},
+			want:     true,
+			reported: true,
+			wantMsg:  "\nunexpected Close call for failed WebSocket connection",
+		},
+		{
+			name: "websocket is closed",
+			fields: fields{
+				conn:     &websocket.Conn{},
+				isClosed: true,
+			},
+			args: args{
+				reporter: newStringReporter(),
+				where:    "Close",
+			},
+			want:     true,
+			reported: true,
+			wantMsg:  "\nunexpected Close call for closed WebSocket connection",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Websocket{
+				chain:    makeChain(tt.args.reporter),
+				conn:     tt.fields.conn,
+				isClosed: tt.fields.isClosed,
+			}
+			if got := c.checkUnusable(tt.args.where); got != tt.want {
+				t.Errorf("Websocket.checkUnusable() = %v, want %v", got, tt.want)
+				return
+			}
+			if got := tt.args.reporter.reported; got != tt.reported {
+				t.Errorf("Websocket.checkUnusable() is error message reported = %v, want %v",
+					got, tt.want)
+				return
+			}
+			if got := tt.args.reporter.msg; got != tt.wantMsg {
+				t.Errorf("Websocket.checkUnusable() error message = %v, want %v", got, tt.want)
+				return
+			}
+		})
+	}
+}
+
+func TestWebsocketWriteJSONMarshalFail(t *testing.T) {
```

Comment:
Here and in tests below we check failures or different methods: WriteJSON, WriteMessage, etc.

I suggest to add a single successful case for each method too.

So we, for example, first check that WriteJSON() works with a our good setup, then change exactly one necessary thing in our setup a bit to make it bad, and check that WriteJSON() fails on it.

This approach allows to ensure that we're testing something meaningful even if we don't check the exact error messages. We had a good setup, it worked, changed one thing, it stopped working - thus we're really testing something.

Also note that the while successful cases are mostly covered by integration tests, it's handy to cover them with unit tests as well, since a failing unit test is much more informative than a failing integration test, due to its simplicity.

Reply:
Testing good setup for each function almost always requires functional gorilla websocket Conn. Unfortunately, Conn is not an interface so I can't make a mock for it. I would need to create an actual websocket Conn and using the actual websocket Conn would make unit tests similar to integration tests. Should I proceed with adding unit tests which use actual websocket connection?

Reply:
Oh, really, I didn't think of that. But maybe just make the Conn an interface as you mentioned? It seems a mock would be easy to make a mock.

Reply:
I can't convert Conn to interface since it is defined in github.com/gorilla/websocket import :(

Reply:
We can just create our own interface, say, WesocketConn, with the subset of gorilla/websocket.Conn methods that we're actually using. Then we should use this interface instead of gorilla/websocket.Conn in websocket.go.

Response.Websocket() will pass a real gorilla conn as the interface implementation. However our websocket unit tests will pass a mock implementation.

Reply:
Great idea! I have added WebsocketConn interface and success unit test cases in recent commits.